vllm/distributed/device_communicators/shm_broadcast.py

```diff
@@ -0,0 +1,145 @@
+                if written_flag and read_count != self.world_size - 1:
+                    # this block is written and not read by all readers
+                    # try to write to the next block
+                    self.current_idx = (self.current_idx + 1) % self.max_chunks
+                    if self.current_idx == start_index:
+                        # no empty block found
+                        # wait for a while (0.1 us)
+                        time.sleep(1e-7)
+                    continue
+                # found a block that is either
+                # (1) not written
+                # (2) read by all readers
+                # let caller write to the buffer
+                with self.data as buf:
+                    yield buf
+
+                # caller has written to the buffer
+                # reset the state
+                metadata_buffer[0] = 1
+                for i in range(1, self.world_size):
+                    metadata_buffer[i] = 0
+                break
+
+    @contextmanager
+    def acquire_read(self):
+        assert self.is_reader, "Only readers can acquire read"
+        start_index = self.current_idx
+        while True:
+            with self.metadata as metadata_buffer:
+                read_flag = metadata_buffer[self.rank]
+                written_flag = metadata_buffer[0]
+                if not written_flag or read_flag:
+                    # this block is either
+                    # (1) not written
+                    # (2) already read by this reader
+                    # try to read the next block
+                    self.current_idx = (self.current_idx + 1) % self.max_chunks
+                    if self.current_idx == start_index:
+                        # no block found
+                        # wait for a while (0.1 us)
+                        time.sleep(1e-7)
+                    continue
+                # found a block that is not read by this reader
+                # let caller read from the buffer
+                with self.data as buf:
+                    yield buf
+
+                # caller has read from the buffer
+                # set the read flag
+                metadata_buffer[self.rank] = 1
+                break
+
+    def broadcast_object(self, obj=None):
+        if self.is_writer:
+            serialized_obj = pickle.dumps(obj,
+                                          protocol=pickle.HIGHEST_PROTOCOL)
+            if len(serialized_obj) > self.max_chunk_bytes:
+                raise RuntimeError(
+                    f"{len(serialized_obj)=} larger than the allowed value "
+                    f"{self.max_chunk_bytes},"
+                    "Please increase the max_chunk_bytes parameter.")
+            with self.acquire_write() as buf:
+                buf[:len(serialized_obj)] = serialized_obj
+            return obj
+        else:
+            with self.acquire_read() as buf:
+                obj = pickle.loads(buf)
```

Comment:
Out of curiosity, why we don't need to know the serialized object size here?

Reply:
see https://github.com/vllm-project/vllm/issues/4440#issuecomment-2081674870 . the pickle serialization format is self-ended.

Reply:
Thanks! It's better to add a comment to avoid confusion from others in the future.

Reply:
fixed in [b8105bb](https://github.com/vllm-project/vllm/pull/5399/commits/b8105bb72684805c9fd422a7454802fe5d76a67f)